feat(rss): implement RSS feed for spotlight entries

prestons18 opened the pull request:
![Passing RSS validity](https://github.com/user-attachments/assets/d0f563d8-5f4f-4c3a-9f45-6baa1c9b83c2)
![RSS XML sample](https://github.com/user-attachments/assets/e50c8332-7638-46ca-846c-8e9716a1a748)

(refer to comments on rss.xml.ts for more information)
closes #125

prestons18 commented:
Looks good to push, as minimal.
![image](https://github.com/user-attachments/assets/c75c67da-d736-4646-a712-d2908e29e3df)
![image](https://github.com/user-attachments/assets/c96fb1b4-ad40-4950-ab73-359a0ce51f5a)

One issue: Like I mentioned about images, Astro produces weird URLs. Even when escaped. XML does not allow this,
or I may just be doing something wrong haha. It's logging the customData, so you could investigate this behavior. Currently, it's not being appended to the XML for the validity error.
![image](https://github.com/user-attachments/assets/caef7fb9-1494-4467-89c5-3a239e205239)

pawelgrzybek commented on src/pages/rss.xml.ts:
```diff
@@ -0,0 +1,100 @@
+
+const baseUrl = 'https://nn1.dev';
+
+let cachedXML: string | null = null;
```
We may not need this one.

pawelgrzybek commented on src/pages/rss.xml.ts:
```diff
@@ -0,0 +1,100 @@
+type SpotlightEntry = CollectionEntry<"spotlight">;
+
+export const GET: APIRoute = async () => {
+  if (cachedXML) {
```
We can remove this one and rely on the Netlify CDN caching controlled by headers. We already have Netlify adapter connected to the project. I think adding single header, can do the job for us. I'm not sure about the values, but look like day of caching and week for revalidation is ok starting point. Do you agree?

```js
    return new Response(cachedXML, {
       headers: {
          "Content-Type": "application/rss+xml; charset=utf-8",
          "Netlify-CDN-Cache-Control": "public, durable, s-maxage=86400, stale-while-revalidate=604800"
        },
     });


```

pawelgrzybek commented on src/pages/rss.xml.ts:
```diff
@@ -0,0 +1,100 @@
+    const spotlights: SpotlightEntry[] = await getCollection("spotlight");
+
+    // Filter out future-dated posts
+    const publishedSpotlights = spotlights.filter(spotlight =>
```
We can simplify this one by filtering out all of the spotlights that are in the draft mode.

```
spotlights.filter(s => !s.draft)
```

Something like that should do.

pawelgrzybek reviewed (commented):
Such a good work dude. We need to align on few little details before we can merge it. I hope you will be ok with my suggestions.

pawelgrzybek commented on src/pages/rss.xml.ts:
```diff
@@ -0,0 +1,100 @@
+// A fast, cached build-time only RSS.
```
This one is a nit-picky, but what do you think about changing the name of this one to `spotlight.xml`? The `rss.xml` is a very generic one, and may lock us in the future. Maybe at some point we can have a blog, or notes, or events or some other entities that also expose a feed. By being more specific we can avoid some conflicts in the future.

Also, we need to link to this file in the html `head` section.

```
<link rel="alternate" type="application/rss+xml" title="NN1 Dev Club" href="https://nn1.dev/spotlight.xml">
```

prestons18 commented:
Appreciate the detailed review, @pawelgrzybek !

You've raised some great points, and I've implemented all of them, other than content output.

Reason being: @astrojs/rss requires XML namespaces for handling content:encoded, and it automatically infers it. I've tried several approaches to add the namespace normally & manually, but none have successfully appended it.

I've also cleaned up the file for a more organised approach, please do let me know if you have any further suggestions.

One more thing, I'll need to implement a proper caching mechanism because Netlify's caching headers doesn't impact the serverless function’s internal runtime cache (declared as cachedXML). Restarting the server to refresh the cache definitely isn’t ideal!

prestons18 commented:
As I've mentioned, I ran into a few issues with @astrojs/rss, so I ended up building my own RSS library instead. This has helped reduce the LOC in spotlight.xml.ts and gave me more control, and now we have: content, intelligent descriptions, and enclosures (images).

I still need to test enclosures thoroughly in a real hosted environment, and I'll also handle replacing images in the content to ensure URLs are valid.

Been a good little run tonight, will improve code & push all.

![image](https://github.com/user-attachments/assets/d6ca0984-0ec9-4af1-9aa4-d9940176bdbc)

pawelgrzybek commented:
You did a golden work mate, really fantastic work and I cannot thank you enough. I did a proper review of it and I think this is good state to land on prod man. We can refine that later, but this is an incredible work you did. Screenshot below from my reader.

<img width="1305" alt="Screenshot 2025-06-25 at 07 13 45" src="https://github.com/user-attachments/assets/2cbd7e68-c61f-4e33-9d0f-576b262e4dbf" />

pawelgrzybek reviewed (approved):
Freakin legend 🙇‍♂️

prestons18 commented:
Glad this is resolved! Pleasure working with you
